contrib/unit_test_coverage/generate_coverage_report.sh

```diff
@@ -0,0 +1,133 @@
+#!/bin/bash
+#
+# Copyright (c) 2020 The Orbit Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# --- generate_coverage_report.sh ---
+#
+# This script generates a coverage report for the Orbit repository. It uses 
+# llvm-cov and requires a build with build arguments -fprofile-instr-generate 
+# and -fcoverage-mapping. 
+#
+# usage: generate_coverage_report.sh SOURCE_DIR BUILD_DIR OUTPUT_DIR
+#
+# settings:
+# grep inverted matching list for files in BUILD_DIR/bin/ that are not used
+BIN_FILE_FILTER="\.debug\|crashpad_handler"
+# grep inverted matching list for source files that are not used
+SOURCE_FILE_FILTER="Test.cpp\|Tests.cpp\|Fuzzer.cpp"
+# ------------------------------------
+
+# fail on any error
+set -e
+
+function get_css_class_string {
+  local percent=$1
+  local result="column-entry-red"
+  if (( $(echo "$percent > 80" | bc) )); then
+    result="column-entry-yellow"
+  fi
+  if (( $(echo "$percent >= 100" | bc) )); then
+    result="column-entry-green"
+  fi
+  echo "$result"
+}
+
+function generate_html_td {
+  local count=$1
+  local covered=$2
+  local percent=$(jq -n "$covered / $count * 100")
+  local percent_rounded=$(printf "%0.2f" $percent)
+  local css_class=$(get_css_class_string $percent_rounded)
+  local percent_formatted=$(printf "%6s" "$percent_rounded")
+  echo "<td class='$css_class'><pre> $percent_formatted% ($covered/$count)</pre></td>"
+}
+
+if [ "$#" -ne 3 ]; then
+  echo "error: wrong number of arguments."
+  echo "usage: generate_coverage_report.sh SOURCE_DIR BUILD_DIR OUTPUT_DIR"
+  exit 1
+fi
+
+SOURCE_DIR=$(realpath $1)
+if [ ! -d $SOURCE_DIR ]; then
+  echo "error: directory $1 does not exist."
+  exit 1
+fi
+
+BUILD_DIR=$(realpath $2)
+if [ ! -d $BUILD_DIR ]; then
+  echo "error: directory $2 does not exist."
+  exit 1
+fi
+
+OUTPUT_DIR=$(realpath $3)
+if [ ! -d $OUTPUT_DIR ]; then
+  echo "error: directory $3 does not exist."
+  exit 1
+fi
+
+LIBS=$BUILD_DIR/lib/*
+BINS=$(find $BUILD_DIR/bin/ -maxdepth 1 -type f | grep -v "$BIN_FILE_FILTER")
+FORMATTED_BINARY_FILES=$(echo $LIBS $BINS | sed "s/ / -object=/g")
+SUM_FUNCTIONS_COUNT=0
+SUM_FUNCTIONS_COVERED=0
+SUM_LINES_COUNT=0
+SUM_LINES_COVERED=0
+SUM_REGIONS_COUNT=0
+SUM_REGIONS_COVERED=0
+
+for COMPONENT in $(cd "$BUILD_DIR"; ctest --show-only=json-v1 | jq -r '.tests | .[] | .name')
```

Comment:
Personally, I'd appreciated a brief comment, what you are iterating here exactly. In particular the piped part. Maybe also move the "cd" out of the loop.

Reply:
I added a comment to clarify, but I like the `cd` as part of this command. 

I am doing the cd, because ctest needs to be run in the build directory. I would prefer to tell ctest the build dir as a parameter, but I was not able to figure out how. I feel the way I did it is the next best thing to that, because it does not actually change the working directory of the script, but only for the subshell this command is executed in.